[WPE] WPEPlatform-based implementation of Interaction Media Features

spenap opened the pull request:
#### 2e472a4b1063f7eb10f34fe875cfab1aa7c83cd5
<pre>
[WPE] WPEPlatform-based implementation of Interaction Media Features
<a href="https://bugs.webkit.org/show_bug.cgi?id=289166">https://bugs.webkit.org/show_bug.cgi?id=289166</a>

Reviewed by Carlos Garcia Campos.

This updates the WebPageWPE behaviour so that it queries
interaction-related information (i.e. mouse or touch capabilities) from
the WebProcess singleton. The WebProcess retrieves this information
throug the WebProcessCreationParameters, which themselves are fetched by
means of WPEPlatform.

This initial changeset includes the Wayland implementation where seat
capabilities are queried, then stored as properties. When a change is
received by the seat, it is propagated to the WebProcesses.

The old implementation has been kept as a fallback: when not using
WPE_PLATFORM and using TOUCH_EVENTS, Touch is assumed. Otherwise we
fallback to Mouse.

* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/wpe/PlatformScreenWPE.cpp:
(WebCore::screenHasTouchDevice): Deleted.
(WebCore::screenIsTouchPrimaryInputDevice): Deleted.
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/AvailableInputDevices.h: Added.
* Source/WebKit/Shared/glib/AvailableInputDevices.serialization.in: Added.
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::toAvailableInputDevices):
(WebKit::WebProcessPool::availableInputDevices):
(WebKit::WebProcessPool::platformInitialize):
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::platformInvalidateContext):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpeDisplaySetProperty):
(wpeDisplayGetProperty):
(wpe_display_class_init):
(wpe_display_get_available_input_devices):
(wpe_display_set_available_input_devices):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandSetup):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.cpp:
(WPE::WaylandSeat::availableInputDevices const):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.h:
(WPE::WaylandSeat::setAvailableInputDevicesChangedCallback):
* Source/WebKit/WebProcess/WebPage/wpe/WebPageWPE.cpp:
(WebKit::WebPage::hoverSupportedByPrimaryPointingDevice const):
(WebKit::WebPage::hoverSupportedByAnyAvailablePointingDevice const):
(WebKit::WebPage::pointerCharacteristicsOfPrimaryPointingDevice const):
(WebKit::WebPage::pointerCharacteristicsOfAllAvailablePointingDevices const):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::primaryPointingDevice const):
(WebKit::WebProcess::setAvailableInputDevices):

Canonical link: <a href="https://commits.webkit.org/293276@main">https://commits.webkit.org/293276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/add3ff824b54dca9c7fbc217a58ca5e0e187c5dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17958 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103447 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48856 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26409 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74845 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32011 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13823 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88804 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55205 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/97813 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13606 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6765 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48298 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83586 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105821 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25414 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18494 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83826 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85006 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83296 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21057 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27938 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5610 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19058 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30553 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25192 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/PlatformWPE.cmake:
```diff
@@ -120,6 +120,8 @@ list(APPEND WebKit_SERIALIZATION_IN_FILES
 
     Shared/soup/WebCoreArgumentCodersSoup.serialization.in
+
+    Shared/wpe/AvailableInputTypes.serialization.in
```
This patch is for now wpe specific, but we could use the same approach for gtk, so I would add this in a glib dir instead of wpe. Also `InputTypes` can be confusing name, since we use the same for the form inputs, see `InputType.h`. I would use AvailableInputDevices instead, or something similar.

carlosgcampos commented on Source/WebKit/Shared/wpe/AvailableInputTypes.h:
```diff
@@ -0,0 +1,36 @@
+ */
+
+#pragma once
+
```
This should be guarded with `#if PLATFORM(WPE)`

carlosgcampos commented on Source/WebKit/Shared/wpe/AvailableInputTypes.h:
```diff
@@ -0,0 +1,36 @@
+enum class AvailableInputTypes : uint8_t {
+    Mouse       = (1 << 0),
+    Keyboard    = (1 << 1),
+    Touch       = (1 << 2),
```
TouchScreen? to make it clear this is not about touchpads?

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -133,6 +133,8 @@ void WebProcessPool::platformInitializeWebProcess(const WebProcessProxy& process
         parameters.implementationLibraryName = FileSystem::fileSystemRepresentation(String::fromLatin1(wpe_loader_get_loaded_implementation_library_name()));
     }
+
+    parameters.availableInputTypes = availableInputTypes();
```
In addition to the initial parameter we need an IPC message to be sent to all web processes when the capabilities change.

spenap replied:
I've created an `UpdateAvailableInputDevices` message, let me know what you think

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -302,6 +304,31 @@ const String& WebProcessPool::generateNextAccessibilityBusName()
+#if ENABLE(WPE_PLATFORM)
+    bool usingWPEPlatformAPI = !!g_type_class_peek(WPE_TYPE_DISPLAY);
+    if (!usingWPEPlatformAPI)
+        return AvailableInputTypes::Mouse;
```
This is not correct, when building with wpe platform enabled, but using the old api, we should do the same was when not building with wpe-platform enabled. Old API should always do the same. I would move the new api handling inside `if (usingWPEPlatformAPI)` and returning in thar case. Then instead of `#elif`  use `#endif ` and then return Touch if touch events enabled or Mouse otherwise.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -104,11 +115,59 @@ static void wpeDisplayDispose(GObject* object)
+        if (display->priv->availableInputTypes != g_value_get_flags(value))  {
+            display->priv->availableInputTypes =  static_cast<WPEAvailableInputTypes>(g_value_get_flags(value));
+            g_object_notify_by_pspec(object, paramSpec);
+        }
```
We should add a public setter and call it from here. Just document in the API docs of the setter that it should only be called by platform implementations

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -227,8 +227,14 @@ const struct wl_registry_listener registryListener = {
+            priv->wlSeat = makeUnique<WPE::WaylandSeat>(static_cast<struct wl_seat*>(wl_registry_bind(registry, name, &wl_seat_interface, std::min<uint32_t>(version, 8))), [weakDisplay = GWeakPtr { display }]() {
+                if (GRefPtr<WPEDisplayWayland> protectedDisplay { weakDisplay.get() }) {
+                    auto* priv = protectedDisplay->priv;
+                    g_object_set(protectedDisplay.get(), "available-input-types", priv->wlSeat->availableInputTypes(), nullptr);
```
And use the setter here

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.cpp:
```diff
@@ -39,9 +39,10 @@ namespace WPE {
+WaylandSeat::WaylandSeat(struct wl_seat* seat, Function<void()>&& callback)
     : m_seat(seat)
     , m_keymap(adoptGRef(wpe_keymap_xkb_new()))
+    , m_capabilitiesChangedCallback(WTFMove(callback))
```
I would add a explicit setter for the callback, in the constructor it's not obvious what the callback is for

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/wpe/WebPageWPE.cpp:
```diff
@@ -48,38 +48,34 @@ bool WebPage::platformCanHandleRequest(const ResourceRequest&)
-    return true;
-#endif
+    // For now, there's no concept of primary vs any
+    return WebProcess::singleton().availableInputTypes().contains(AvailableInputTypes::Mouse);
```
I think we could use a simple approach for now like, if there's a mouse that's the primary.

carlosgcampos commented on Source/WebKit/Shared/glib/AvailableInputDevices.h:
```diff
@@ -0,0 +1,40 @@
+enum class AvailableInputDevices : uint8_t {
+    Mouse       = (1 << 0),
+    Keyboard    = (1 << 1),
+    TouchScreen = (1 << 2),
```
```suggestion
    Touchscreen = (1 << 2),
```

carlosgcampos commented on Source/WebKit/Shared/glib/AvailableInputDevicesPrivate.h:
```diff
@@ -0,0 +1,41 @@
+
+#include "AvailableInputDevices.h"
+
+#include <wpe/wpe-platform.h>
```
This doesn't belong here, this is a file in Shared, but the wpe platform api is a UI process only api.

carlosgcampos commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -751,6 +754,10 @@ class WebProcessPool final
 #endif
 
+#if PLATFORM(WPE)
+    OptionSet<AvailableInputDevices> availableInputDevices();
```
```suggestion
    OptionSet<AvailableInputDevices> availableInputDevices() const;
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -66,6 +66,7 @@
 #endif
 
 #if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
+#include "AvailableInputDevicesPrivate.h"
```
We normally use Private when there's a public header, but that's no the case here.

carlosgcampos commented on Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp:
```diff
@@ -58,6 +73,11 @@ ScreenManager::ScreenManager()
+        auto availableInputDevices = toAvailableInputDevices(wpe_display_get_available_input_devices(d));
+        for (auto& pool : WebProcessPool::allProcessPools())
+            pool->sendToAllProcesses(Messages::WebProcess::UpdateAvailableInputDevices(availableInputDevices));
+    }), this);
```
I don't think this belongs here either. We could monitor the devices from the process pool and then notify all its processes. It's not common to have many process pools, so it's ok to monitor the same property from all of them. That way we don't need a common place for the conversion helper, it can be a static function in WebProcessPoolGLib.cpp. We can connect to the signal in WebProcessPool::platformInitialize

carlosgcampos replied:
And we should probably disconnect the signal too, in WebProcessPool::platformInvalidateContext probably

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -585,3 +641,38 @@ gboolean wpe_display_use_explicit_sync(WPEDisplay* display)
+ * wpe_display_get_available_input_devices:
+ * @display: a #WPEDisplay
+ *
+ * Get availableInputDevices of @display that can be used by creators to adjust their UI based on the available interactions.
```
```suggestion
 * Get the available input devices of @display that can be used by creators to adjust their UI based on the available interactions.
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -585,3 +641,38 @@ gboolean wpe_display_use_explicit_sync(WPEDisplay* display)
+ * @display: a #WPEDisplay
+ * @availableInputDevices: a #WPEAvailableInputDevices
+ *
+ * Sets availableInputDevices for a @display.
```
```suggestion
 * Set the available input devices for @display.
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -585,3 +641,38 @@ gboolean wpe_display_use_explicit_sync(WPEDisplay* display)
+/**
+ * wpe_display_set_available_input_devices:
+ * @display: a #WPEDisplay
+ * @availableInputDevices: a #WPEAvailableInputDevices
```
```suggestion
 * @devices: a #WPEAvailableInputDevices
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
```diff
@@ -85,29 +92,32 @@ typedef enum {
+WPE_API WPESettings             *wpe_display_get_settings                  (WPEDisplay *display);
+WPE_API WPEAvailableInputDevices wpe_display_get_available_input_devices   (WPEDisplay *display);
+WPE_API void                     wpe_display_set_available_input_devices   (WPEDisplay *display,
+                                                                            WPEAvailableInputDevices availableInputDevices);
```
```suggestion
                                                                            WPEAvailableInputDevices devices);
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -227,8 +227,15 @@ const struct wl_registry_listener registryListener = {
-        else if (interfaceName == "wl_seat"_s)
+        else if (interfaceName == "wl_seat"_s) {
             priv->wlSeat = makeUnique<WPE::WaylandSeat>(static_cast<struct wl_seat*>(wl_registry_bind(registry, name, &wl_seat_interface, std::min<uint32_t>(version, 8))));
+            priv->wlSeat->setAvailableInputDevicesChangedCallback([weakDisplay = GWeakPtr { display }]() {
```
Let's do this in wpeDisplayWaylandSetup()

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -227,8 +227,15 @@ const struct wl_registry_listener registryListener = {
+        else if (interfaceName == "wl_seat"_s) {
             priv->wlSeat = makeUnique<WPE::WaylandSeat>(static_cast<struct wl_seat*>(wl_registry_bind(registry, name, &wl_seat_interface, std::min<uint32_t>(version, 8))));
+            priv->wlSeat->setAvailableInputDevicesChangedCallback([weakDisplay = GWeakPtr { display }]() {
+                if (GRefPtr<WPEDisplayWayland> protectedDisplay { weakDisplay.get() }) {
```
GWeakPtr is not thread safe and wpe-platform api either, so here we don't need to protect the display, we can just check the weak and get the pointer. So I would early return and then just use weakDisplay.get() directly.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -227,8 +227,15 @@ const struct wl_registry_listener registryListener = {
+            priv->wlSeat->setAvailableInputDevicesChangedCallback([weakDisplay = GWeakPtr { display }]() {
+                if (GRefPtr<WPEDisplayWayland> protectedDisplay { weakDisplay.get() }) {
+                    auto* priv = protectedDisplay->priv;
+                    wpe_display_set_available_input_devices(WPE_DISPLAY(protectedDisplay.get()), priv->wlSeat->availableInputDevices());
```
Since we are keeping the callback approahc in the end, we could just pass the new value and here we don't need to ask the seat.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.h:
```diff
@@ -119,6 +123,7 @@ class WaylandSeat {
         GWeakPtr<WPEToplevelWayland> toplevel;
         HashMap<int32_t, std::pair<double, double>, IntHash<int32_t>, WTF::SignedWithZeroKeyHashTraits<int32_t>> points;
     } m_touch;
+    Function<void ()> m_capabilitiesChangedCallback;
```
```suggestion
    Function<void()> m_capabilitiesChangedCallback;
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/wpe/WebPageWPE.cpp:
```diff
@@ -48,38 +48,34 @@ bool WebPage::platformCanHandleRequest(const ResourceRequest&)
-    return true;
-#endif
+    // For now, there's no concept of primary vs any
+    return WebProcess::singleton().availableInputDevices().contains(AvailableInputDevices::Mouse);
```
We can add WebProcess::primaryPointingDevice() and for now just return mouse if present, otherwise touchscreen if available or none

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/wpe/WebPageWPE.cpp:
```diff
@@ -48,38 +48,34 @@ bool WebPage::platformCanHandleRequest(const ResourceRequest&)
+    const auto& availableInputs = WebProcess::singleton().availableInputDevices();
+    if (availableInputs.contains(AvailableInputDevices::Mouse))
+        return PointerCharacteristics::Fine;
+    if (availableInputs.contains(AvailableInputDevices::TouchScreen))
```
Which is wat you are doing here, but this could check WebProcess::primaryPointingDevice() instead and the logic to device the primary is in one place.

carlosgcampos commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -126,6 +126,9 @@ messages -> WebProcess : AuxiliaryProcess WantsAsyncDispatchMessage {
     SetScreenProperties(struct WebCore::ScreenProperties screenProperties)
 #endif
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
+    UpdateAvailableInputDevices(OptionSet<WebKit::AvailableInputDevices> availableInputDevices)
```
```suggestion
    SetAvailableInputDevices(OptionSet<WebKit::AvailableInputDevices> availableInputDevices)
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -102,6 +117,18 @@ void WebProcessPool::platformInitialize(NeedsGlobalStaticInitialization)
+        g_signal_connect(display, "notify::available-input-devices", G_CALLBACK(+[](WPEDisplay* d, GParamSpec*, WebProcessPool*) {
+            auto availableInputDevices = toAvailableInputDevices(wpe_display_get_available_input_devices(d));
+            for (auto& pool : allProcessPools())
+                pool->sendToAllProcesses(Messages::WebProcess::SetAvailableInputDevices(availableInputDevices));
```
This is done from every pool, only send the message to the processes of this pool.

spenap replied:
oops! I borrowed this bit from some other patch, and didn't realise what I was doing, thanks for spotting it!

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -102,6 +117,18 @@ void WebProcessPool::platformInitialize(NeedsGlobalStaticInitialization)
         installMemoryPressureHandler();
 #endif
+
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
```
```suggestion
#if ENABLE(WPE_PLATFORM)
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -102,6 +117,18 @@ void WebProcessPool::platformInitialize(NeedsGlobalStaticInitialization)
+    bool usingWPEPlatformAPI = !!g_type_class_peek(WPE_TYPE_DISPLAY);
+    if (usingWPEPlatformAPI) {
+        auto* display = wpe_display_get_primary();
+        g_signal_connect(display, "notify::available-input-devices", G_CALLBACK(+[](WPEDisplay* d, GParamSpec*, WebProcessPool*) {
```
```suggestion
        g_signal_connect(display, "notify::available-input-devices", G_CALLBACK(+[](WPEDisplay* display, GParamSpec*, WebProcessPool*) {
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -185,6 +214,13 @@ void WebProcessPool::platformInitializeWebProcess(const WebProcessProxy& process
 
 void WebProcessPool::platformInvalidateContext()
 {
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
```
```suggestion
#if ENABLE(WPE_PLATFORM)
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -302,6 +338,24 @@ const String& WebProcessPool::generateNextAccessibilityBusName()
 }
 
+#if PLATFORM(WPE)
+OptionSet<AvailableInputDevices> WebProcessPool::availableInputDevices() const
```
This could be a static function and then we don't need to add platform specific code to WebProcessPool.h

carlosgcampos commented on Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
```diff
@@ -297,6 +299,25 @@ void WebProcess::setScreenProperties(const WebCore::ScreenProperties& properties
+#if PLATFORM(WPE)
+std::optional<AvailableInputDevices> WebProcess::primaryPointingDevice() const
+{
+    const auto& availableInputs = availableInputDevices();
```
Why not just use m_availableInputDevices?

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -77,6 +77,39 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+#endif
+
+#if PLATFORM(WPE)
+static OptionSet<AvailableInputDevices> WebProcessPool::availableInputDevices()
```
```suggestion
static OptionSet<AvailableInputDevices> availableInputDevices()
```

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -102,6 +135,17 @@ void WebProcessPool::platformInitialize(NeedsGlobalStaticInitialization)
+    bool usingWPEPlatformAPI = !!g_type_class_peek(WPE_TYPE_DISPLAY);
+    if (usingWPEPlatformAPI) {
+        auto* display = wpe_display_get_primary();
+        g_signal_connect(display, "notify::available-input-devices", G_CALLBACK(+[](WPEDisplay* display, GParamSpec*, WebProcessPool*) {
```
```suggestion
        g_signal_connect(display, "notify::available-input-devices", G_CALLBACK(+[](WPEDisplay* display, GParamSpec*, WebProcessPool* pool) {
```

spenap committed:
[WPE] WPEPlatform-based implementation of Interaction Media Features
https://bugs.webkit.org/show_bug.cgi?id=289166

Reviewed by Carlos Garcia Campos.

This updates the WebPageWPE behaviour so that it queries
interaction-related information (i.e. mouse or touch capabilities) from
the WebProcess singleton. The WebProcess retrieves this information
throug the WebProcessCreationParameters, which themselves are fetched by
means of WPEPlatform.

This initial changeset includes the Wayland implementation where seat
capabilities are queried, then stored as properties. When a change is
received by the seat, it is propagated to the WebProcesses.

The old implementation has been kept as a fallback: when not using
WPE_PLATFORM and using TOUCH_EVENTS, Touch is assumed. Otherwise we
fallback to Mouse.

* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/wpe/PlatformScreenWPE.cpp:
(WebCore::screenHasTouchDevice): Deleted.
(WebCore::screenIsTouchPrimaryInputDevice): Deleted.
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/AvailableInputDevices.h: Added.
* Source/WebKit/Shared/glib/AvailableInputDevices.serialization.in: Added.
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::toAvailableInputDevices):
(WebKit::WebProcessPool::availableInputDevices):
(WebKit::WebProcessPool::platformInitialize):
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::platformInvalidateContext):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpeDisplaySetProperty):
(wpeDisplayGetProperty):
(wpe_display_class_init):
(wpe_display_get_available_input_devices):
(wpe_display_set_available_input_devices):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandSetup):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.cpp:
(WPE::WaylandSeat::availableInputDevices const):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEWaylandSeat.h:
(WPE::WaylandSeat::setAvailableInputDevicesChangedCallback):
* Source/WebKit/WebProcess/WebPage/wpe/WebPageWPE.cpp:
(WebKit::WebPage::hoverSupportedByPrimaryPointingDevice const):
(WebKit::WebPage::hoverSupportedByAnyAvailablePointingDevice const):
(WebKit::WebPage::pointerCharacteristicsOfPrimaryPointingDevice const):
(WebKit::WebPage::pointerCharacteristicsOfAllAvailablePointingDevices const):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::primaryPointingDevice const):
(WebKit::WebProcess::setAvailableInputDevices):

Canonical link: https://commits.webkit.org/293276@main
